execution/executor/src/components/apply_chunk_output.rs

```diff
@@ -0,0 +1,272 @@
+                if !o.write_set().is_empty() || !o.events().is_empty() {
+                    error!(
+                        "Discarded transaction has non-empty write set or events. \
+                     Transaction: {:?}. Status: {:?}.",
+                        t,
+                        o.status(),
+                    );
+                    DIEM_EXECUTOR_ERRORS.inc();
+                }
+                Ok(t)
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok((new_epoch_marker.is_some(), to_keep, to_discard, to_retry))
+    }
+
+    fn apply_write_set(
+        state_cache: StateCache,
+        new_epoch: bool,
+        to_keep: &[(Transaction, TransactionOutput)],
+    ) -> Result<(
+        Vec<HashMap<AccountAddress, AccountStateBlob>>,
+        Vec<(HashValue, HashMap<NibblePath, HashValue>)>,
+        SparseMerkleTree<AccountStateBlob>,
+        Option<EpochState>,
+    )> {
+        let StateCache {
+            frozen_base,
+            mut accounts,
+            proofs,
+        } = state_cache;
+
+        // Apply write sets to account states in the AccountCache, resulting in new account states.
+        let account_states = to_keep
+            .iter()
+            .map(|(t, o)| process_write_set(t, &mut accounts, o.write_set().clone()))
+            .collect::<Result<Vec<_>>>()?;
+        let account_blobs = account_states
+            .par_iter()
+            .with_min_len(100)
+            .map(|account_to_state| {
+                account_to_state
+                    .iter()
+                    .map(|(addr, state)| Ok((*addr, AccountStateBlob::try_from(state)?)))
+                    .collect::<Result<HashMap<_, _>>>()
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        // apply new account states to the base state tree, resulting in updated state tree.
+        let (roots_with_node_hashes, result_state) = frozen_base
+            .serial_update(
+                Self::account_blobs_to_smt_updates(&account_blobs),
+                &ProofReader::new(proofs),
+            )
+            .map_err(|e| anyhow!("Failed to update state tree. err: {:?}", e))?;
+        let result_state = result_state.unfreeze();
+
+        // Get the updated validator set from updated account state.
+        let next_epoch_state = if new_epoch {
+            Some(Self::parse_validator_set(&accounts)?)
+        } else {
+            None
+        };
+
+        Ok((
+            account_blobs,
+            roots_with_node_hashes,
+            result_state,
+            next_epoch_state,
+        ))
+    }
+
+    fn account_blobs_to_smt_updates(
+        account_blobs: &[HashMap<AccountAddress, AccountStateBlob>],
+    ) -> Vec<Vec<(HashValue, &AccountStateBlob)>> {
+        account_blobs
+            .iter()
+            .map(|m| {
+                m.iter()
+                    .map(|(account, blob)| (HashAccountAddress::hash(account), blob))
```

Comment:
account.hash()?

Reply:
There's also CryptoHash being imported in this file.